### DISCUSSION TITLE
Best practice to search for file names recursively respecting .gitignore?

### DISCUSSION THREAD

**Discussion by rolfb** [BurntSushi/ripgrep#1589] [General] [answered]
When searching for file names recursively like this when I want to find all index.js files:  
  
`rg --files **/index.js`

Which flags can I set to make it respect my .gitignore which ignores folders like the `node_modules/` folder?

Using ripgrep v12.1.0 on macOS Catalina.

**Comment by BurntSushi** [comment, marked as answer]
You cannot cause explicit positional parameters like you've provided to be ignored by `.gitignore`. This is an intentional design decision. In particular, ripgrep does not see `**/index.js`. It sees the _expansion_ of `**/index.js` to every individual matching file by your shell. Therefore, ripgrep cannot differentiate between `rg foo/index.js` and `rg **/index.js`. They are literally exactly the same. Thus, if someone ran `rg foo/index.js` and `foo/index.js` was in their `.ignore` and ripgrep respected ignore files for positional parameters, then `rg foo/index.js` would not search anything. This would be highly unexpected, but does unfortunately lead to these sorts of cases where you want that ignore filtering to happen since you're providing globs.

The way to work around this is to express your globs not to your shell, but to ripgrep itself. Here's an example:

```
$ mkdir -p foo/bar/baz
$ echo test > foo/index.js
$ echo test > foo/bar/index.js
$ echo test > foo/bar/baz/index.js
$ mkdir -p node_modules/quux
$ echo test > node_modules/quux/index.js

$ rg -l test
foo/index.js
foo/bar/baz/index.js
foo/bar/index.js
node_modules/quux/index.js

$ echo node_modules > .ignore

$ rg -l test
foo/bar/baz/index.js
foo/bar/index.js
foo/index.js

$ rg -l test **/index.js
node_modules/quux/index.js
foo/index.js
foo/bar/index.js
foo/bar/baz/index.js

$ rg -l test -g 'index.js'
foo/bar/baz/index.js
foo/bar/index.js
foo/index.js
```

The last example is the key. It says, "search the current directory recursively, with smart filtering, but only search files named `index.js`." It includes everything in `foo` but ignores `node_modules` since `node_modules` is in our `.ignore` file.

**Reply by rolfb** [reply]
Thanks, that's exactly what I wanted to know. Cheers!
